File: PS/백준 단계별 문제풀기/백트래킹/9663nqueen.cpp
#include<iostream>
#include<vector>
#include<math.h>
using namespace std;


int n;
int loc[15];
int result = 0;

bool check(int cnt) {
    bool isPromising = true;

    //열이같은열인지
    //대각선인지
    for (int i = 0;i < cnt;i++) {
        if (loc[i] == loc[cnt] || abs(loc[cnt]-loc[i]) == abs(cnt - i)) {
            isPromising = false;
            break;
        }
    }
    return isPromising;
}


void backtracking(int cnt) {
    if (cnt == n) {
        result++;
        return;
    }

    for (int i =0;i <n;i++) {
        loc[cnt] = i;
        if (check(cnt)) {
            backtracking(cnt + 1);
        }
    }

}



int main() {
    cin >> n;
    backtracking(0);
    cout << result << "\n";
    return 0;
}

File: PS/백준 단계별 문제풀기/백트래킹/9663 N-Queen.py
n = int(input())
queens = [-1]*n
ans = 0
visited = [False]*n


def isPromising(r, c):
    for i in range(r):
        if queens[i] == c or abs(r-i) == abs(c-queens[i]):
            return False
    return True


def backtracking(count):
    global ans
    if count == n:
        ans += 1
        return

    for i in range(n):
        if not visited[i] and isPromising(count, i):
            queens[count] = i
            visited[i] = True
            backtracking(count+1)
            visited[i] = False


backtracking(0)
print(ans)
